File: leetcode/Python/0055-m-jump-game.py
"""
DP
Time: O(n^2) - TLE
Space: O(n)
"""
class Solution_1:
    def canJump(self, nums: List[int]) -> bool:
        n = len(nums)
        
        # init
        f = [False] * n
        f[0] = True

        # function
        for curt in range(1, n):
            for prev in range(curt):
                if f[prev] and prev + nums[prev] >= curt:
                    f[curt] = True
                    break
        # answer
        return f[n - 1]

"""
Greedy
"""
class Solution_2:
    def canJump(self, nums: List[int]) -> bool:
        lastPos = len(nums) - 1
        for i in range(len(nums) - 1, -1, -1):
            if i + nums[i] >= lastPos:
                lastPos = i
        return lastPos == 0

File: leetcode/C++/0055-m-jump-game.cpp
// Greedy
// Time: O(n)
// Space: O(1)
class Solution {
public:
    bool canJump(vector<int>& nums) {
        int n = nums.size();
        int next_pos = 0;
        for (int i = 0; i < n; i++) {
            if (i <= next_pos && i + nums[i] >= next_pos)
                next_pos = i + nums[i];
        }
        return next_pos >= n - 1;
    }
};


// DP
// Time: O(n^2)
// Space: O(n)
class Solution {
public:
    bool canJump(vector<int>& nums) {
        int n = nums.size();
        vector<bool> f(n, false);
        
        f[0] = true;
        for (int j = 1; j < n; j++) {
            for (int i = 0; i < j; i++) {
                if (f[i] && i + nums[i] >= j) {
                    f[j] = true;
                    break;
                }
            }
        }
        
        return f[n - 1];
    }
};